### PR TITLE
Resolve `PHP_INT_MAX` as positive integer

### PR DESCRIPTION
**Pull request by herndlm**
Fixes https://github.com/phpstan/phpstan/issues/5657

This is my first PR, I definitely don't know enough about phpstan internals and might have missed something.

I was no sure what this should resolve to. It would be enough and best to set it to "positive int" but I think this concept is not existing, is it? I then chose the positive range because
1. Hard-coding it to `PHP_INT_MAX` feels not correct as that would be the value of the platform phpstan is running on
2. Setting it to something like 1 could make even more problems
3. Setting it to the UnionType of what the most common values for 32 and 64 bit might make problems (e.g. the 64 bit code will not work on 32 bit I assume?)

Initially I tried to be smart and dynamically resolve all int constant via `constant` but realised that this a bad idea because of the already mentioned 1. What do you think?

### CODE REVIEW

**Comment by herndlm** [comment]
Ok well, looks like I broke things anyway :) I'll look into that.

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -1978,6 +1978,9 @@ private function resolveType(Expr $node): Type
 				if ($resolvedConstantName === '__COMPILER_HALT_OFFSET__') {
 					return new IntegerType();
 				}
+				if ($resolvedConstantName === 'PHP_INT_MAX') {
+					return new IntegerRangeType(1, null);
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:1982]: Hi, the comment in IntegerRangeType's constructor says: "this constructor can be made private when PHP 7.2 is the minimum"

Everyone should use `IntegerRangeType::fromInterval()` instead.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
This PR is mostly fine, thank you :) Also please note there are other constants that would benefit from the same treatment, for example `PHP_INT_MIN` is definitely one of them but there are many more.

I'd appreciate if you could go through this list and created more specific types: https://github.com/phpstan/phpstan-src/blob/43c192b23f4159beb122d07ec8b4e2d84628236e/conf/config.neon#L113-L157

Integers can be made more specific using IntegerRangeType, strings can have `AccessoryNonEmptyStringType` and `AccessoryNumericStringType` attached to them in `IntersectionType`.

I'd like if you'd find some time to do it in the next PR after this one is merged. Thank you!

**Comment by herndlm** [comment]
> This PR is mostly fine, thank you :) Also please note there are other constants that would benefit from the same treatment, for example `PHP_INT_MIN` is definitely one of them but there are many more.
> 
> I'd appreciate if you could go through this list and created more specific types:
> 
> https://github.com/phpstan/phpstan-src/blob/43c192b23f4159beb122d07ec8b4e2d84628236e/conf/config.neon#L113-L157
> 
> Integers can be made more specific using IntegerRangeType, strings can have `AccessoryNonEmptyStringType` and `AccessoryNumericStringType` attached to them in `IntersectionType`.
> 
> I'd like if you'd find some time to do it in the next PR after this one is merged. Thank you!

Right, I thought about that, but wanted to keep it simple and small here. How do you prefer the PRs for those? E.g. all in one, or grouped by affected area or type. It depends I guess..

Also: I tried to add `PHP_INT_MIN` too with the same test case basically but using `PHP_INT_MIN` instead of `PHP_INT_MAX` which should then fail. Interestingly phpstan seemed to end up in an endless loop then, or I did no wait long enough, but how would you debug that?

File: src/Analyser/MutatingScope.php (hunk 1 above)
Comment (herndlm) [reply src/Analyser/MutatingScope.php]: Out-of-scope: but wouldn't it make sense to add `@deprecated` to the constructor then?

**Comment by ondrejmirtes** [comment]
The other constants should be tested using NodeScopeResolverTest and the `assertType` function. You can group them together in the next PR after this one is merged :) 

> how would you debug that?

No idea why that would be happening. You should isolate the problem to a smallest file possible, run that with `bin/phpstan analyse` separately and then go through the code and figure out where the infinite run comes from.

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**